improver/lightning.py

```diff
@@ -204,3 +205,215 @@ def latitude_to_threshold(
+    Outputs:
+    20 km lightning probability over the valid time of the accumulated precipitation
+
+
+    Description of the algorithm:
+
+    Regression equation when CAPE and APCP are greater than zero:
+    lprob= cape*APCP
+    lprob=0.13*alog(lprob+0.7)+0.05
+
+    If APCP is very low, a separate regression equation is used to predict lightning probability.
+    The definition of “very low” is raised slightly when PWAT values are high. This is because
+    the model often produces showery precipitation that doesn’t access the actual instability in
+    very moist environments:
+    lprob_noprecip=CAPE/(CIN+100.0)
+    lprob_noprecip=0.025*alog(lprob_noprecip+0.31)+0.03
+    APCP=APCP-(PWAT/1000)
+
+    IF APCP is less than 0.01 THEN lprob=lprob_noprecip
+    If there is no CAPE but the atmosphere is “close” to unstable, lightning does sometimes occur,
+    especially when heavy precipitation may have stabilized the atmosphere in the model. Unstable
+    values of lifted index are positive here:
+    LIFTIDX=LIFTIDX+4.0
+    IF LIFTIDX is less than 0 THEN LIFTIDX=0.0
+    IF CAPE is less than 0 THEN lprob =0.2*(LIFTIDX*APCP)^0.5
+
+    Finally, the probability of lightning is reduced when there is not much PW, because graupel
+    cannot form and start the whole charging process. Therefore we reduce the probability:
+    IF PWAT is less than 20 THEN lprob=lprob*(PWAT/20.0)
+
+    Limit final probabilities to 95% as that is as skillful as the regression equations could get:
+    IF lprob is greater than 0.95 THEN lprob=0.95
+    """
+
+    @staticmethod
+    def _extract_input(cubes: CubeList, cube_name: str) -> Cube:
+        """Extract the relevant cube based on the cube name.
+
+        Args:
+            cubes: Cubes from which to extract required input.
+            cube_name: Name of cube to extract.
+
+        Returns:
+            The extracted cube.
+        """
+        try:
+            cube = cubes.extract_cube(iris.Constraint(cube_name))
+        except iris.exceptions.ConstraintMismatchError:
+            raise ValueError(f"No cube named {cube_name} found in {cubes}")
+        return cube
+
+    def _get_inputs(self, cubes: CubeList) -> Tuple[Cube, Cube]:
+        """
+        Separates CAPE, LI, PW, CIN, and APCP cubes and checks that the following
+        match: forecast_reference_time, spatial coords, time-bound interval and
+        that CAPE time is at the lower bound of precipitation accumulation time.
+        """
+
+        output_cubes = iris.cube.CubeList()
+        input_names = [
+            "atmosphere_specific_convective_available_potential_energy",
+            "temperature_difference_between_ambient_air_and_air_lifted_adiabatically",
+            "precipitable_water",
+            "atmosphere_specific_convective_inhibition",
+            "precipitation_amount",
+        ]
+
+        for input_name in input_names:
+            output_cubes.append(self._extract_input(cubes, input_name))
```

Comment:
```suggestion
        input_names = {
            "atmosphere_specific_convective_available_potential_energy": ["J kg-1"],
            "temperature_difference_between_ambient_air_and_air_lifted_adiabatically": ["K"],
            "precipitable_water": ["kg m-2", "mm"],
            "atmosphere_specific_convective_inhibition": ["J kg-1"],
            "precipitation_amount": ["kg m-2", "mm"],
        }

        for input_name, units in input_names.items():
            output_cubes.append(self._extract_input(cubes, input_name))
            if not output_cubes[-1].units in units:
                raise ValueError("Input diagnostic units are incorrect")
```

Can I suggest something along these lines to check that the inputs units are what we expect, and if not raise an exception? It seems particularly important here as the units are stripped later on so we need to ensure they're correct whilst we can.

Reply:
This would probably want a better exception message than my rather vague suggestion.